From c55d9569b3f006783a66897bd6d7d16702fefeb0 Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Tue, 28 Mar 2023 16:19:06 +0530
Subject: [PATCH] ath12k: Fix incorrect value of rx_per_chain_rssi

Currently while using PRINT_ARRAY_TO_BUF to print values
of rx_per_chain_rssi_ext_in_dbm, values are currupted
because rx_per_chain_rssi_ext_in_dbm variable requires %d
to be printed correctly.

Using CHAIN_ARRAY_TO_BUF macro to print
rx_per_chain_rssi_ext_in_dbm.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 .../net/wireless/ath/ath12k/debugfs_htt_stats.c  | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
index 01d7f67..fc4cbb9 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs_htt_stats.c
@@ -3595,6 +3595,17 @@ fail:
 		kfree(rx_gi[i]);
 }
 
+#define CHAIN_ARRAY_TO_BUF(out, buflen, arr, len)                         \
+	do {                                       \
+		int index = 0; u8 i;                           \
+		for (i = 0; i < len; i++) {                    \
+			index += scnprintf((out + buflen) + index,         \
+			(ATH12K_HTT_STATS_BUF_SIZE - buflen) - index,  \
+				" %u:%d,", i, arr[i]);           \
+		}                                  \
+		buflen += index;                           \
+	} while (0)
+
 static inline void htt_print_rx_pdev_rate_ext_stats_tlv(const void *tag_buf,
 						       struct debug_htt_stats_req *stats_req)
 {
@@ -3625,9 +3636,10 @@ static inline void htt_print_rx_pdev_rate_ext_stats_tlv(const void *tag_buf,
 	for (j = 0; j < HTT_RX_PDEV_STATS_NUM_SPATIAL_STREAMS; j++) {
 		len += scnprintf(buf + len, buf_len - len,
 				 "\nrx_per_chain_rssi_ext_in_dbm[%u] = ", j);
-		PRINT_ARRAY_TO_BUF(buf, len,
+		CHAIN_ARRAY_TO_BUF(buf, len,
 				   htt_stats_buf->rx_per_chain_rssi_ext_in_dbm[j],
-				   NULL, HTT_RX_PDEV_STATS_NUM_BW_EXT_COUNTERS, "\n");
+				   HTT_RX_PDEV_STATS_NUM_BW_EXT_COUNTERS);
+		len += scnprintf(buf + len, buf_len - len, "\n");
 	}
 
 	PRINT_ARRAY_TO_BUF(buf, len, htt_stats_buf->rx_mcs_ext,
-- 
2.17.1

